tag task fails in multi-module project

```
$ ./gradlew :spec:tag

> Task :spec:tag FAILED

FAILURE: Build failed with an exception.

* What went wrong:
Execution failed for task ':spec:tag'.
> org.eclipse.jgit.errors.RepositoryNotFoundException: repository not found: /home/user/Projects/personal/personal/Semver.kt/spec
```

--- a/Makefile
+++ b/Makefile
@@ -31,8 +31,6 @@ test: ## Runs tests for the project
 local: _check_java ## Publishes artifacts to local repos
 	# publish to local maven
 	./gradlew publishToMavenLocal publishAllPublicationsToLocalPluginRepoRepository -Pversion=0.0.0
-	# push git tag
-	git push origin --tags
 
 .PHONY: release
 release: _check_java test ## Publishes the next release

--- a/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
+++ b/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
@@ -14,6 +14,7 @@ import org.eclipse.jgit.api.Git
 import org.gradle.testkit.runner.TaskOutcome
 import org.gradle.util.GradleVersion
 import java.nio.file.Path
+import kotlin.io.path.createDirectories
 import kotlin.io.path.createFile
 import kotlin.io.path.writeText
 
@@ -296,6 +297,40 @@ class SemverKtPluginFT : DescribeSpec({
 
             // re-enable soft assertions
             assertSoftly = false
+        }
+    }
+
+    describe("multi-module project") {
+        it("should set next version via commit message") {
+            val project = SemverKtTestProject()
+            // arrange
+            Git.open(project.projectDir.toFile()).use {
+                it.tag().setName("v0.1.0").call() // set initial version
+                project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                it.add().addFilepattern("text.txt").call()
+                it.commit().setMessage("New commit\n\n[minor]").call()
+            }
+            // act
+            val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun"))
+            // assert
+            result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
+            result.output shouldContain "Calculated next version: 0.2.0"
+        }
+
+        it("should set next version via gradle property") {
+            val project = SemverKtTestProject()
+            // arrange
+            Git.open(project.projectDir.toFile()).use {
+                it.tag().setName("v0.1.0").call() // set initial version
+                project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                it.add().addFilepattern("text.txt").call()
+                it.commit().setMessage("New commit").call()
+            }
+            // act
+            val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun", "-Pincrement=minor"))
+            // assert
+            result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
+            result.output shouldContain "Calculated next version: 0.2.0"
         }
     }
 })

--- a/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
+++ b/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
@@ -184,48 +184,93 @@ class SemverKtPluginFT : DescribeSpec({
             // assert
             result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
             // initial version should be calculated from config, so the keyword value doesn't really matter so long as it's valid
-            result.output shouldContain "> Configure project :\nProject version: 0.0.0"
+            result.output shouldContain "> Configure project :\nProject test-project version: 0.0.0"
             result.output shouldContain "> Task :tag\nNot doing anything"
             result.output shouldNotContain "Calculated next version"
         }
     }
 
     describe("current git HEAD with a version") {
-        it("should set version to current version with increment from commit message") {
-            val project = SemverKtTestProject()
-            // arrange
-            Git.open(project.projectDir.toFile()).use {
-                project.projectDir.resolve("text.txt").createFile().writeText("Hello")
-                it.add().addFilepattern("text.txt").call()
-                it.commit().setMessage("New commit\n[minor]").call()
-                it.tag().setName("v0.1.0").call() // add a tag manually on latest commit
+        listOf(true, false).forEach { dryRun ->
+            it("should set version to current version with increment from commit message${if (dryRun) " and dryRun" else ""}") {
+                val project = SemverKtTestProject()
+                // arrange
+                Git.open(project.projectDir.toFile()).use {
+                    project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                    it.add().addFilepattern("text.txt").call()
+                    it.commit().setMessage("New commit\n[minor]").call()
+                    it.tag().setName("v0.1.0").call() // add a tag manually on latest commit
+                }
+                // act
+                val args = if (dryRun) arrayOf("tag", "-PdryRun") else arrayOf("tag")
+                val result = Builder.build(project = project, args = args)
+                // assert
+                result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
+                // initial version should be calculated from config, so the keyword value doesn't really matter so long as it's valid
+                result.output shouldContain "> Configure project :\nProject test-project version: 0.1.0"
+                result.output shouldContain "> Task :tag\nCurrent version: 0.1.0"
+                result.output shouldNotContain "Calculated next version"
             }
-            // act
-            val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun"))
-            // assert
-            result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
-            // initial version should be calculated from config, so the keyword value doesn't really matter so long as it's valid
-            result.output shouldContain "> Configure project :\nProject version: 0.1.0"
-            result.output shouldContain "> Task :tag\nCurrent version: 0.1.0"
-            result.output shouldNotContain "Calculated next version"
+            it("should set version to current version with increment from gradle properties${if (dryRun) " and dryRun" else ""}") {
+                val project = SemverKtTestProject()
+                // arrange
+                Git.open(project.projectDir.toFile()).use {
+                    project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                    it.add().addFilepattern("text.txt").call()
+                    it.commit().setMessage("New commit").call()
+                    it.tag().setName("v0.1.0").call() // add a tag manually on latest commit
+                }
+                // act
+                val args = if (dryRun) arrayOf("tag", "-PdryRun", "-Pincrement=minor") else arrayOf("tag", "-Pincrement=minor")
+                val result = Builder.build(project = project, args = args)
+                // assert
+                result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
+                // initial version should be calculated from config, so the keyword value doesn't really matter so long as it's valid
+                result.output shouldContain "> Configure project :\nProject test-project version: 0.1.0"
+                result.output shouldContain "> Task :tag\nCurrent version: 0.1.0"
+                result.output shouldNotContain "Calculated next version"
+            }
         }
-        it("should set version to current version with increment from gradle properties") {
-            val project = SemverKtTestProject()
-            // arrange
-            Git.open(project.projectDir.toFile()).use {
-                project.projectDir.resolve("text.txt").createFile().writeText("Hello")
-                it.add().addFilepattern("text.txt").call()
-                it.commit().setMessage("New commit").call()
-                it.tag().setName("v0.1.0").call() // add a tag manually on latest commit
+    }
+
+    describe("re-run 'tag' after releasing a version") {
+        listOf(true, false).forEach { dryRun ->
+            it("should return UP_TO_DATE with increment from commit${if (dryRun) " and dryRun" else ""}") {
+                val project = SemverKtTestProject()
+                // arrange
+                Git.open(project.projectDir.toFile()).use {
+                    project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                    it.add().addFilepattern("text.txt").call()
+                    it.commit().setMessage("New commit\n\n[minor]").call()
+                }
+                Builder.build(project = project, args = arrayOf("tag", "-Pincrement=minor")) // release a version
+                // act
+                val args = if (dryRun) arrayOf("tag", "-PdryRun") else arrayOf("tag")
+                val result = Builder.build(project = project, args = args)
+                // assert
+                result.task(":tag")?.outcome shouldBe TaskOutcome.UP_TO_DATE
+                result.output shouldContain "> Configure project :\nProject test-project version: 0.1.0"
+                result.output shouldContain "> Task :tag UP-TO-DATE"
+                result.output shouldNotContain "Calculated next version"
             }
-            // act
-            val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun", "-Pincrement=minor"))
-            // assert
-            result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
-            // initial version should be calculated from config, so the keyword value doesn't really matter so long as it's valid
-            result.output shouldContain "> Configure project :\nProject version: 0.1.0"
-            result.output shouldContain "> Task :tag\nCurrent version: 0.1.0"
-            result.output shouldNotContain "Calculated next version"
+            it("should return UP_TO_DATE with increment from gradle properties${if (dryRun) " and dryRun" else ""}") {
+                val project = SemverKtTestProject()
+                // arrange
+                Git.open(project.projectDir.toFile()).use {
+                    project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                    it.add().addFilepattern("text.txt").call()
+                    it.commit().setMessage("New commit").call()
+                }
+                Builder.build(project = project, args = arrayOf("tag", "-Pincrement=minor")) // release a version
+                // act
+                val args = if (dryRun) arrayOf("tag", "-PdryRun", "-Pincrement=minor") else arrayOf("tag", "-Pincrement=minor")
+                val result = Builder.build(project = project, args = args)
+                // assert
+                result.task(":tag")?.outcome shouldBe TaskOutcome.UP_TO_DATE
+                result.output shouldContain "> Configure project :\nProject test-project version: 0.1.0"
+                result.output shouldContain "> Task :tag UP-TO-DATE"
+                result.output shouldNotContain "Calculated next version"
+            }
         }
     }
 

--- a/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
+++ b/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
@@ -243,7 +243,7 @@ class SemverKtPluginFT : DescribeSpec({
                     it.add().addFilepattern("text.txt").call()
                     it.commit().setMessage("New commit\n\n[minor]").call()
                 }
-                Builder.build(project = project, args = arrayOf("tag", "-Pincrement=minor")) // release a version
+                Builder.build(project = project, args = arrayOf("tag")) // release a version
                 // act
                 val args = if (dryRun) arrayOf("tag", "-PdryRun") else arrayOf("tag")
                 val result = Builder.build(project = project, args = args)
@@ -346,48 +346,54 @@ class SemverKtPluginFT : DescribeSpec({
     }
 
     describe("multi-module project") {
-        it("should set next version via commit message") {
-            val project = SemverKtTestProject(multiModule = true)
-            // arrange
-            Git.open(project.projectDir.toFile()).use {
-                it.tag().setName("v0.1.0").call() // set initial version
-                project.projectDir.resolve("text.txt").createFile().writeText("Hello")
-                it.add().addFilepattern("text.txt").call()
-                it.commit().setMessage("New commit\n\n[minor]").call()
+        listOf(true, false).forEach { dryRun ->
+            it("should set next version via commit${if (dryRun) " and dryRun" else ""}") {
+                val project = SemverKtTestProject(multiModule = true)
+                // arrange
+                Git.open(project.projectDir.toFile()).use {
+                    it.tag().setName("v0.1.0").call() // set initial version
+                    project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                    it.add().addFilepattern("text.txt").call()
+                    it.commit().setMessage("New commit\n\n[minor]").call()
+                }
+                // act
+                val args = if (dryRun) arrayOf("tag", "-PdryRun") else arrayOf("tag")
+                val result = Builder.build(project = project, args = args)
+                // assert
+                result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
+                result.output shouldContain """
+                    > Task :tag
+                    Calculated next version: 0.2.0
+
+                    > Task :submodule:tag
+                    Calculated next version: 0.2.0
+                    ${if (!dryRun) "Tag v0.2.0 already exists in project" else ""}
+                """.trimIndent().trim()
             }
-            // act
-            val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun"))
-            // assert
-            result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
-            result.output shouldContain """
-                > Task :tag
-                Calculated next version: 0.2.0
 
-                > Task :submodule:tag
-                Calculated next version: 0.2.0
-            """.trimIndent()
-        }
+            it("should set next version via gradle property${if (dryRun) " and dryRun" else ""}") {
+                val project = SemverKtTestProject(multiModule = true)
+                // arrange
+                Git.open(project.projectDir.toFile()).use {
+                    it.tag().setName("v0.1.0").call() // set initial version
+                    project.projectDir.resolve("text.txt").createFile().writeText("Hello")
+                    it.add().addFilepattern("text.txt").call()
+                    it.commit().setMessage("New commit").call()
+                }
+                // act
+                val args = if (dryRun) arrayOf("tag", "-PdryRun", "-Pincrement=minor") else arrayOf("tag", "-Pincrement=minor")
+                val result = Builder.build(project = project, args = args)
+                // assert
+                result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
+                result.output shouldContain """
+                    > Task :tag
+                    Calculated next version: 0.2.0
 
-        it("should set next version via gradle property") {
-            val project = SemverKtTestProject(multiModule = true)
-            // arrange
-            Git.open(project.projectDir.toFile()).use {
-                it.tag().setName("v0.1.0").call() // set initial version
-                project.projectDir.resolve("text.txt").createFile().writeText("Hello")
-                it.add().addFilepattern("text.txt").call()
-                it.commit().setMessage("New commit").call()
+                    > Task :submodule:tag
+                    Calculated next version: 0.2.0
+                    ${if (!dryRun) "Tag v0.2.0 already exists in project" else ""}
+                """.trimIndent().trim()
             }
-            // act
-            val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun", "-Pincrement=minor"))
-            // assert
-            result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
-            result.output shouldContain """
-                > Task :tag
-                Calculated next version: 0.2.0
-
-                > Task :submodule:tag
-                Calculated next version: 0.2.0
-            """.trimIndent()
         }
     }
 })

--- a/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
+++ b/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPluginFT.kt
@@ -277,7 +277,7 @@ class SemverKtPluginFT : DescribeSpec({
                     // just add some dumb assertion to verify gradle actually ran successfully
                 ).also { it.task(":clean")?.outcome shouldBe TaskOutcome.UP_TO_DATE }
                 // - add plugin to settings.gradle file
-                project.writePluginSettings()
+                project.writePluginSettings(false)
                 // - add a file and commit with release keyword
                 Git.open(project.projectDir.toFile()).use {
                     project.projectDir.resolve("text.txt").createFile().writeText("Hello")
@@ -302,7 +302,7 @@ class SemverKtPluginFT : DescribeSpec({
 
     describe("multi-module project") {
         it("should set next version via commit message") {
-            val project = SemverKtTestProject()
+            val project = SemverKtTestProject(multiModule = true)
             // arrange
             Git.open(project.projectDir.toFile()).use {
                 it.tag().setName("v0.1.0").call() // set initial version
@@ -314,11 +314,17 @@ class SemverKtPluginFT : DescribeSpec({
             val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun"))
             // assert
             result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
-            result.output shouldContain "Calculated next version: 0.2.0"
+            result.output shouldContain """
+                > Task :tag
+                Calculated next version: 0.2.0
+
+                > Task :submodule:tag
+                Calculated next version: 0.2.0
+            """.trimIndent()
         }
 
         it("should set next version via gradle property") {
-            val project = SemverKtTestProject()
+            val project = SemverKtTestProject(multiModule = true)
             // arrange
             Git.open(project.projectDir.toFile()).use {
                 it.tag().setName("v0.1.0").call() // set initial version
@@ -330,7 +336,13 @@ class SemverKtPluginFT : DescribeSpec({
             val result = Builder.build(project = project, args = arrayOf("tag", "-PdryRun", "-Pincrement=minor"))
             // assert
             result.task(":tag")?.outcome shouldBe TaskOutcome.SUCCESS
-            result.output shouldContain "Calculated next version: 0.2.0"
+            result.output shouldContain """
+                > Task :tag
+                Calculated next version: 0.2.0
+
+                > Task :submodule:tag
+                Calculated next version: 0.2.0
+            """.trimIndent()
         }
     }
 })

--- a/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/fixture/SemverKtTestProject.kt
+++ b/semantic-versioning/src/functionalTest/kotlin/io/github/serpro69/semverkt/gradle/plugin/fixture/SemverKtTestProject.kt
@@ -4,10 +4,12 @@ import io.github.serpro69.semverkt.release.configuration.Configuration
 import org.eclipse.jgit.api.Git
 import java.nio.file.Path
 import kotlin.io.path.absolutePathString
+import kotlin.io.path.createDirectories
 import kotlin.io.path.writeText
 
 class SemverKtTestProject(
-    defaultSettings: Boolean = false
+    defaultSettings: Boolean = false,
+    multiModule: Boolean = false,
 ) : AbstractProject() {
 
     private val gradlePropertiesFile = projectDir.resolve("gradle.properties")
@@ -15,11 +17,48 @@ class SemverKtTestProject(
     private val buildFile = projectDir.resolve("build.gradle.kts")
     private val configFile = projectDir.resolve("semantic-versioning.json")
 
-    constructor(defaultSettings: Boolean = false, configure: (dir: Path) -> Configuration) : this(defaultSettings) {
+    constructor(
+        defaultSettings: Boolean = false,
+        multiModule: Boolean = false,
+        configure: (dir: Path) -> Configuration,
+    ) : this(defaultSettings, multiModule) {
         configFile.writeText(configure(projectDir).jsonString())
     }
 
-    fun writePluginSettings() {
+    init {
+        gradlePropertiesFile.writeText(
+            """
+            org.gradle.jvmargs=-Dfile.encoding=UTF-8
+            version=0.0.0
+            """.trimIndent()
+        )
+
+        if (defaultSettings) settingsFile.writeText(
+            """
+            rootProject.name = "test-project"
+
+            ${if (multiModule) "include(\"submodule\")" else ""}
+            """.trimIndent()
+        )
+        else writePluginSettings(multiModule)
+
+        // Apply our plugin
+        buildFile.writeBuildFile()
+
+        if (multiModule) projectDir.resolve("submodule").createDirectories().also {
+            it.resolve("build.gradle.kts").writeText("""
+                plugins {
+                }
+            """.trimIndent())
+        }
+
+        Git.open(projectDir.toFile()).use {
+            it.add().addFilepattern(".").call()
+            it.commit().setMessage("add project files").call()
+        }
+    }
+
+    fun writePluginSettings(multiModule: Boolean) {
         settingsFile.writeText(
             """
             import java.nio.file.Paths
@@ -38,48 +77,20 @@ class SemverKtTestProject(
                     }
                 }
             }
+
+            ${if (multiModule) "include(\"submodule\")" else ""}
             """.trimIndent()
         )
     }
 
-    init {
-//        projectDir.resolve("abc.txt").writeText("hello world")
-//
-//        projectDir.resolve("sub").createDirectories()
-//            .also {
-//                it.resolve("build.gradle.kts")
-//                .writeText("""
-//                    plugins {
-//                    }
-//                """.trimIndent())
-//
-//                it.resolve("abc.txt")
-//                    .writeText("another one")
-//            }
-        // Yes, this is independent of our plugin project's properties file
-        gradlePropertiesFile.writeText(
-            """
-            org.gradle.jvmargs=-Dfile.encoding=UTF-8
-            version=0.0.0
-            """.trimIndent()
-        )
-
-        // Yes, our project under test can use build scans. It's a real project!
-        if (defaultSettings) settingsFile.writeText(
-            """
-            rootProject.name = "test-project"
-            """.trimIndent()
-        )
-        else writePluginSettings()
-
-        // Apply our plugin
-        buildFile.writeText(
+    private fun Path.writeBuildFile() {
+        writeText(
             """
             plugins {
                 java
             }
       
-            println("Project version: ${'$'}{project.version}")
+            println("Project ${'$'}{project.name} version: ${'$'}{project.version}")
 
             tasks.create("ft", Test::class.java) {
                 doFirst {
@@ -88,10 +99,5 @@ class SemverKtTestProject(
             }
             """.trimIndent()
         )
-
-        Git.open(projectDir.toFile()).use {
-            it.add().addFilepattern(".").call()
-            it.commit().setMessage("add project files").call()
-        }
     }
 }

--- a/semantic-versioning/src/main/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPlugin.kt
+++ b/semantic-versioning/src/main/kotlin/io/github/serpro69/semverkt/gradle/plugin/SemverKtPlugin.kt
@@ -2,15 +2,12 @@ package io.github.serpro69.semverkt.gradle.plugin
 
 import io.github.serpro69.semverkt.gradle.plugin.tasks.TagTask
 import io.github.serpro69.semverkt.release.Increment
-import io.github.serpro69.semverkt.release.Increment.DEFAULT
-import io.github.serpro69.semverkt.release.Increment.MAJOR
-import io.github.serpro69.semverkt.release.Increment.MINOR
 import io.github.serpro69.semverkt.release.Increment.NONE
-import io.github.serpro69.semverkt.release.Increment.PATCH
-import io.github.serpro69.semverkt.release.Increment.PRE_RELEASE
 import io.github.serpro69.semverkt.release.SemverRelease
 import io.github.serpro69.semverkt.release.configuration.JsonConfiguration
+import io.github.serpro69.semverkt.spec.Semver
 import org.gradle.api.Plugin
+import org.gradle.api.Project
 import org.gradle.api.initialization.Settings
 import org.gradle.api.logging.LogLevel
 import org.gradle.api.logging.Logging
@@ -29,97 +26,109 @@ class SemverKtPlugin : Plugin<Settings> {
         }
         // override configuration via settings extension
         settings.extensions.create("semantic-versioning", SemverPluginExtension::class.java, config)
-        // configure allprojects with semver
-        settings.gradle.allprojects { project ->
-            val propPromoteRelease = project.hasProperty("promoteRelease")
-            val propPreRelease = project.hasProperty("preRelease")
-            val propIncrement = project.findProperty("increment")
-                ?.let { Increment.getByName(it.toString()) }
-                ?: NONE
 
-            val (currentVersion, latestVersion, nextVersion) = with(SemverRelease(config)) {
-                // IF git HEAD points at a version, set and return it and don't do anything else
-                currentVersion()?.let {
-                    project.version = it
-                    return@with Triple(it, null, null)
+        logger.log(LogLevel.DEBUG, "Using configuration: ${config.jsonString()}")
+
+        // configure all projects with semver
+        settings.gradle.allprojects { project ->
+            configureProject(project, config)
+        }
+
+        logger.log(LogLevel.INFO, "Finish applying plugin...")
+    }
+
+    private fun configureProject(project: Project, config: SemverKtPluginConfig) {
+        val (currentVersion, latestVersion, nextVersion) = setVersion(project, config)
+
+        project.tasks.register("tag", TagTask::class.java) {
+            it.description = "Create a tag for the next version"
+            it.config.set(config)
+            it.dryRun.set(project.hasProperty("dryRun"))
+            // set versions
+            it.currentVersion.set(currentVersion)
+            it.latestVersion.set(latestVersion)
+            it.nextVersion.set(nextVersion)
+        }
+    }
+
+    private fun setVersion(project: Project, config: SemverKtPluginConfig): Triple<Semver?, Semver?, Semver?> {
+        val propPromoteRelease = project.hasProperty("promoteRelease")
+        val propPreRelease = project.hasProperty("preRelease")
+        val propIncrement = project.findProperty("increment")
+            ?.let { Increment.getByName(it.toString()) }
+            ?: NONE
+
+        return with(SemverRelease(config)) {
+            // IF git HEAD points at a version, set and return it and don't do anything else
+            currentVersion()?.let {
+                project.version = it
+                return@with Triple(it, null, null)
+            }
+            // ELSE figure out next version
+            val latestVersion = latestVersion()
+            logger.log(LogLevel.INFO, "Current version: $latestVersion")
+            val increaseVersion = if (propPromoteRelease) NONE else with(nextIncrement()) {
+                logger.log(LogLevel.DEBUG, "Next increment from property: $propIncrement")
+                logger.log(LogLevel.DEBUG, "Next increment from git commit: $this")
+                when (propIncrement) {
+                    // 1. check allowed values for gradle-property based increment
+                    // IF nextIncrement is NONE - we don't want to override it
+                    //  -> it contains some logic for checking existing versions and HEAD version
+                    // ELSE return increment from property
+                    // 2. just return the result of nextIncrement
+                    //  -> DEFAULT and NONE are not valid for property-based increment
+                    !in listOf(
+                        Increment.DEFAULT,
+                        NONE
+                    ) -> if (this == NONE) this else propIncrement
+                    else -> this
                 }
-                // ELSE figure out next version
-                val latestVersion = latestVersion()
-                logger.log(LogLevel.INFO, "Current version: $latestVersion")
-                val increaseVersion = if (propPromoteRelease) NONE else with(nextIncrement()) {
-                    logger.log(LogLevel.DEBUG, "Next increment from property: $propIncrement")
-                    logger.log(LogLevel.DEBUG, "Next increment from git commit: $this")
-                    when (propIncrement) {
-                        // 1. check allowed values for gradle-property based increment
-                        // IF nextIncrement is NONE - we don't want to override it
-                        //  -> it contains some logic for checking existing versions and HEAD version
-                        // ELSE return increment from property
-                        // 2. just return the result of nextIncrement
-                        //  -> DEFAULT and NONE are not valid for property-based increment
-                        !in listOf(DEFAULT, NONE) -> if (this == NONE) this else propIncrement
-                        else -> this
-                    }
+            }
+            logger.log(LogLevel.INFO, "Calculated version increment: '$increaseVersion'")
+            val nextVersion = if (propPromoteRelease) {
+                logger.log(LogLevel.INFO, "Promote to release...")
+                promoteToRelease()
+            } else if (propPreRelease) {
+                logger.log(LogLevel.INFO, "Create pre-release...")
+                createPreRelease(increaseVersion)
+            } else when (increaseVersion) {
+                Increment.MAJOR, Increment.MINOR, Increment.PATCH -> {
+                    logger.log(LogLevel.INFO, "Create release...")
+                    release(increaseVersion)
                 }
-                logger.log(LogLevel.INFO, "Calculated version increment: '$increaseVersion'")
-                val nextVersion = if (propPromoteRelease) {
-                    logger.log(LogLevel.INFO, "Promote to release...")
-                    promoteToRelease()
-                } else if (propPreRelease) {
-                    logger.log(LogLevel.INFO, "Create pre-release...")
-                    createPreRelease(increaseVersion)
-                } else when (increaseVersion) {
-                    MAJOR, MINOR, PATCH -> {
-                        logger.log(LogLevel.INFO, "Create release...")
+                Increment.PRE_RELEASE -> {
+                    latestVersion?.preRelease?.let {
+                        logger.log(LogLevel.INFO, "Next pre-release...")
                         release(increaseVersion)
+                    } ?: run {
+                        logger.log(LogLevel.INFO, "Create default pre-release...")
+                        createPreRelease(Increment.DEFAULT)
                     }
-                    PRE_RELEASE -> {
-                        latestVersion?.preRelease?.let {
-                            logger.log(LogLevel.INFO, "Next pre-release...")
-                            release(increaseVersion)
-                        } ?: run {
-                            logger.log(LogLevel.INFO, "Create default pre-release...")
-                            createPreRelease(DEFAULT)
-                        }
-                    }
-                    DEFAULT, NONE -> null
                 }
-                logger.log(LogLevel.INFO, "Next version: $nextVersion")
-                when {
-                    (nextVersion != null && latestVersion != null) && (nextVersion >= latestVersion) -> {
-                        // only set version if nextVersion >= latestVersion
-                        project.version = nextVersion
-                        logger.log(LogLevel.DEBUG, "Set project.version: ${(project.version)}")
-                    }
-                    nextVersion != null && latestVersion == null -> {
-                        // set initial version
-                        project.version = nextVersion
-                        logger.log(LogLevel.DEBUG, "Set project.version: ${(project.version)}")
-                    }
-                    else -> {
+                Increment.DEFAULT, NONE -> null
+            }
+            logger.log(LogLevel.INFO, "Next version: $nextVersion")
+            when {
+                (nextVersion != null && latestVersion != null) && (nextVersion >= latestVersion) -> {
+                    // only set version if nextVersion >= latestVersion
+                    project.version = nextVersion
+                    logger.log(LogLevel.DEBUG, "Set project.version: ${(project.version)}")
+                }
+                nextVersion != null && latestVersion == null -> {
+                    // set initial version
+                    project.version = nextVersion
+                    logger.log(LogLevel.DEBUG, "Set project.version: ${(project.version)}")
+                }
+                else -> {
 //                    if (config.useSnapshots) {
 //                        project.version = snapshot(increaseVersion)
 //                        logger.log(LogLevel.DEBUG, "Set project.version: ${(project.version)}")
 //                    }
-                        logger.log(
-                            LogLevel.DEBUG,
-                            "Not doing anything: nextVersion $nextVersion < currentVersion $currentVersion"
-                        )
-                    }
+                    logger.log(LogLevel.DEBUG, "Not doing anything...")
                 }
-                logger.log(LogLevel.INFO, "Done...")
-                Triple(null, latestVersion, nextVersion)
             }
-
-            project.tasks.register("tag", TagTask::class.java) {
-                it.description = "Create a tag for the next version"
-                it.currentVersion.set(currentVersion)
-                it.latestVersion.set(latestVersion)
-                it.nextVersion.set(nextVersion)
-                it.dryRun.set(project.hasProperty("dryRun"))
-            }
+            logger.log(LogLevel.INFO, "Done...")
+            Triple(null, latestVersion, nextVersion)
         }
-
-        logger.log(LogLevel.DEBUG, "Using configuration: ${config.jsonString()}")
-        logger.log(LogLevel.INFO, "Finish applying plugin...")
     }
 }

--- a/semantic-versioning/src/main/kotlin/io/github/serpro69/semverkt/gradle/plugin/tasks/TagTask.kt
+++ b/semantic-versioning/src/main/kotlin/io/github/serpro69/semverkt/gradle/plugin/tasks/TagTask.kt
@@ -1,8 +1,12 @@
 package io.github.serpro69.semverkt.gradle.plugin.tasks
 
+import io.github.serpro69.semverkt.gradle.plugin.SemverKtPluginConfig
+import io.github.serpro69.semverkt.release.repo.GitRepository
 import io.github.serpro69.semverkt.spec.Semver
 import org.eclipse.jgit.api.Git
+import org.eclipse.jgit.storage.file.FileRepositoryBuilder
 import org.gradle.api.logging.LogLevel.DEBUG
+import org.gradle.api.logging.LogLevel.INFO
 import org.gradle.api.logging.LogLevel.LIFECYCLE
 import org.gradle.api.logging.Logging
 import org.gradle.api.provider.Property
@@ -15,8 +19,20 @@ abstract class TagTask : SemverReleaseTask() {
     @get:Internal
     internal abstract val dryRun: Property<Boolean>
 
+    @get:Internal
+    internal abstract val config: Property<SemverKtPluginConfig>
+
+    init {
+        // check if tag already set, if so tell plugin it's UP-TO-DATE
+        // NB! it won't handle multi-module projects because currentVersion will be set to all projects
+        // before the task itself is even evaluated and executed
+        outputs.upToDateWhen { currentVersion.orNull != null }
+        //this@TagTask.onlyIf { currentVersion.orNull == null || nextVersion.orNull != null }
+    }
+
     @TaskAction
     fun tag() {
+        // check if tag already set, if so tell plugin it's UP-TO-DATE
         val (current, latest, next) = Triple(currentVersion.orNull, latestVersion.orNull, nextVersion.orNull)
         logger.log(DEBUG, "latestVersion: $latest")
         logger.log(DEBUG, "nextVersion: $next")
@@ -26,29 +42,42 @@ abstract class TagTask : SemverReleaseTask() {
             // release new version
             (next != null && latest != null) && (next > latest) -> {
                 logger.log(LIFECYCLE, "Calculated next version: $next")
-                setTag(next)
+                setTag(next, config.get())
             }
             // release first version
             next != null && latest == null -> {
                 logger.log(LIFECYCLE, "Calculated next version: $next")
-                setTag(next)
+                setTag(next, config.get())
             }
             else -> logger.log(LIFECYCLE, "Not doing anything")
         }
     }
 
-    private fun setTag(nextVer: Semver) {
-        if (!dryRun.get()) with(project.buildFile.parentFile) {
-            logger.log(DEBUG, "Open repo at: $this")
-            val repo = Git.open(this)
-            logger.log(DEBUG, "Set tag to: v$nextVer")
-            repo.setTag("v$nextVer")
+    private fun setTag(nextVer: Semver, config: SemverKtPluginConfig) {
+        if (!dryRun.get()) run {
+            // TODO is there a better way to handle multi-module projects?
+            //  tag might already exist (was set by root project or another sub-module)
+            val isHeadOnTag = GitRepository(config).use { repo ->
+                repo.tags().let { tags ->
+                    tags.isNotEmpty()
+                        && tags.last().let { t -> repo.head() == t.peeledObjectId || repo.head() == t.objectId }
+                }
+            }
+            if (!isHeadOnTag) {
+                logger.log(DEBUG, "Open repo at: $this")
+                FileRepositoryBuilder()
+                    .setWorkTree(project.projectDir)
+                    .findGitDir(project.projectDir)
+                    .build()
+                    .use {
+                        logger.log(DEBUG, "Set tag to: v$nextVer")
+                        Git(it).use { git -> git.setTag("v$nextVer") }
+                    }
+            } else logger.log(INFO, "Tag v$nextVer already exists in project")
         }
     }
 }
 
 private fun Git.setTag(tagName: String) {
-    Git(repository).use { git ->
-        git.tag().setName(tagName).setMessage(tagName).setAnnotated(true).call()
-    }
+    tag().setName(tagName).setMessage(tagName).setAnnotated(true).call()
 }

--- a/semantic-versioning/src/main/kotlin/io/github/serpro69/semverkt/gradle/plugin/tasks/TagTask.kt
+++ b/semantic-versioning/src/main/kotlin/io/github/serpro69/semverkt/gradle/plugin/tasks/TagTask.kt
@@ -55,15 +55,13 @@ abstract class TagTask : SemverReleaseTask() {
 
     private fun setTag(nextVer: Semver, config: SemverKtPluginConfig) {
         if (!dryRun.get()) run {
-            // TODO is there a better way to handle multi-module projects?
-            //  tag might already exist (was set by root project or another sub-module)
-            val isHeadOnTag = GitRepository(config).use { repo ->
-                repo.tags().let { tags ->
-                    tags.isNotEmpty()
-                        && tags.last().let { t -> repo.head() == t.peeledObjectId || repo.head() == t.objectId }
+            // check if tag exists, don't try to create a duplicate
+            val tagExists = GitRepository(config).use { repo ->
+                repo.tags().any { 
+                    Semver(it.name.replace(Regex("""^refs/tags/${config.git.tag.prefix}"""), "")) == nextVer 
                 }
             }
-            if (!isHeadOnTag) {
+            if (!tagExists) {
                 logger.log(DEBUG, "Open repo at: $this")
                 FileRepositoryBuilder()
                     .setWorkTree(project.projectDir)
@@ -73,7 +71,7 @@ abstract class TagTask : SemverReleaseTask() {
                         logger.log(DEBUG, "Set tag to: v$nextVer")
                         Git(it).use { git -> git.setTag("v$nextVer") }
                     }
-            } else logger.log(INFO, "Tag v$nextVer already exists in project")
+            } else logger.log(LIFECYCLE, "Tag v$nextVer already exists in project")
         }
     }
 }

--- a/spec/src/main/kotlin/io/github/serpro69/semverkt/spec/Semver.kt
+++ b/spec/src/main/kotlin/io/github/serpro69/semverkt/spec/Semver.kt
@@ -1,5 +1,6 @@
 package io.github.serpro69.semverkt.spec
 
+import java.io.Serializable
 import java.util.regex.Pattern
 import kotlin.math.sign
 
@@ -24,7 +25,7 @@ fun String.toSemver(prefix: String? = null): Semver {
  *
  * @constructor             creates an instance of [Semver] where semantic version is constructed from the [version] string
  */
-class Semver(private val version: String) : Comparable<Semver> {
+class Semver(private val version: String) : Comparable<Semver>, Serializable {
     private val versionPattern: Pattern =
         Pattern.compile("""^((?:\d+\.){2}\d+)(-(?:[0-9A-Za-z-]*.?)*)?(\+(?:[0-9A-Za-z-]*.?)*)?${'$'}""")
     val normalVersion: String = version.substringBefore("+").substringBefore("-")

